File: agvs_dev/agvs_dev/events/maintenance_visit.py
import frappe
from frappe.utils import getdate, nowdate


def before_submit(doc, method=None):
    if doc.completion_status and not doc.get("custom_verified_by_customer"):
        frappe.throw(
            "Please capture the <b>Verified by Customer</b> signature before submitting."
        )


def on_submit(doc, method=None):
    _update_maintenance_schedule(doc)
    _create_consumed_items_stock_entry(doc)
    _auto_create_delivery_note(doc)


def _update_maintenance_schedule(doc):
    """Update Maintenance Schedule detail rows to Completed on submit."""
    if not doc.maintenance_schedule:
        return

    maint_date = getdate(doc.mntc_date)

    for d in doc.get("purposes") or []:
        if not d.maintenance_schedule_detail:
            continue
        try:
            update_vals = {
                "completion_status": "Completed",
                "actual_date": maint_date
            }
            if d.get("service_person"):
                update_vals["sales_person"] = d.service_person
            frappe.db.set_value(
                "Maintenance Schedule Detail",
                d.maintenance_schedule_detail,
                update_vals
            )
        except Exception:
            frappe.log_error(frappe.get_traceback(), "Maintenance Schedule Detail update failed")

    # Mark whole schedule Completed if all rows done
    try:
        ms = frappe.get_doc("Maintenance Schedule", doc.maintenance_schedule)
        if all(row.completion_status == "Completed" for row in (ms.schedules or [])):
            frappe.db.set_value("Maintenance Schedule", doc.maintenance_schedule, "status", "Completed")
    except Exception:
        frappe.log_error(frappe.get_traceback(), "Maintenance Schedule status update failed")


def _create_consumed_items_stock_entry(doc):
    """Create a Material Issue Stock Entry for additional consumed items on submit."""
    consumed_rows = doc.get("custom_additional_consumed_items") or []
    if not consumed_rows:
        return

    warehouse = doc.get("custom_warehouse")
    if not warehouse:
        frappe.throw(
            "Please set <b>Warehouse</b> in the Additional Consumed Items section "
            "before submitting, as there are items to issue."
        )

    # Skip if already created for this visit
    if frappe.db.exists("Stock Entry", {
        "custom_maintenance_visit": doc.name,
        "stock_entry_type": "Material Issue",
        "docstatus": 1
    }):
        return

    items = []
    for row in consumed_rows:
        if not row.item_code or not (row.consumed_qty or 0) > 0:
            continue
        items.append({
            "item_code": row.item_code,
            "qty": row.consumed_qty,
            "uom": row.uom or "Nos",
            "s_warehouse": warehouse,
            "custom_maintenance_visit": doc.name
        })

    if not items:
        return

    se = frappe.get_doc({
        "doctype": "Stock Entry",
        "stock_entry_type": "Material Issue",
        "company": doc.company,
        "from_warehouse": warehouse,
        "custom_maintenance_visit": doc.name,
        "items": items
    })
    se.insert(ignore_permissions=True)
    se.submit()


def _auto_create_delivery_note(doc):
    """
    Auto-create a Delivery Note (draft) from employee warehouse → client
    if a Material Transfer SE (stores → employee) was created for this MV.
    """
    if not doc.customer:
        return

    # Find the submitted Material Transfer SE linked to this visit
    se_name = frappe.db.get_value(
        "Stock Entry",
        {
            "custom_maintenance_visit": doc.name,
            "stock_entry_type": "Material Transfer",
            "docstatus": 1
        },
        "name"
    )
    if not se_name:
        return

    # Skip if DN already created for this visit
    if frappe.db.exists("Delivery Note", {
        "custom_maintenance_visit": doc.name,
        "docstatus": ["!=", 2]
    }):
        return

    se_doc = frappe.get_doc("Stock Entry", se_name)
    employee_wh = se_doc.to_warehouse  # employee warehouse (target of the MT)

    dn_items = []
    for item in se_doc.items:
        if not item.item_code or not (item.qty or 0) > 0:
            continue
        # Try to get selling rate; fall back to 0
        rate = frappe.db.get_value(
            "Item Price",
            {"item_code": item.item_code, "selling": 1, "currency": frappe.defaults.get_global_default("currency")},
            "price_list_rate"
        ) or 0

        dn_items.append({
            "item_code":   item.item_code,
            "item_name":   item.item_name,
            "qty":         item.qty,
            "uom":         item.uom or "Nos",
            "rate":        rate,
            "warehouse":   employee_wh,
            "description": item.item_name or item.item_code
        })

    if not dn_items:
        return

    try:
        dn = frappe.get_doc({
            "doctype":                 "Delivery Note",
            "customer":                doc.customer,
            "company":                 doc.company,
            "posting_date":            nowdate(),
            "set_warehouse":           employee_wh,
            "custom_maintenance_visit": doc.name,
            "items":                   dn_items
        })
        dn.insert(ignore_permissions=True)
        # Save as draft — user reviews and submits manually
        frappe.msgprint(
            f"Delivery Note <b>{dn.name}</b> created as draft. "
            "Please review and submit it.",
            alert=True
        )
    except Exception:
        frappe.log_error(frappe.get_traceback(), "Auto Delivery Note creation failed")

File: agvs_dev/public/js/maintenance_visit.js
// === Maintenance Visit — Material Transfer button ===

frappe.ui.form.on('Maintenance Visit', {
    refresh(frm) {
        if (frm.is_new()) return;

        // Show "Material Transfer" button only if one hasn't been created yet
        frappe.db.get_list("Stock Entry", {
            filters: {
                custom_maintenance_visit: frm.doc.name,
                stock_entry_type: "Material Transfer",
                docstatus: ["!=", 2]
            },
            fields: ["name"],
            limit: 1
        }).then(res => {
            if (!res || !res.length) {
                frm.add_custom_button(__("Material Transfer"), () => {
                    create_material_transfer(frm);
                }, __("Create"));
            }
        });
    }
});

function create_material_transfer(frm) {
    if (!frm.doc.custom_from_warehouse) {
        frappe.msgprint({
            title: __("Missing Warehouse"),
            indicator: "red",
            message: __("Please set <b>From Warehouse (Stores)</b> before creating Material Transfer.")
        });
        return;
    }

    const purposes = frm.doc.purposes || [];
    const technician_user = purposes.length ? purposes[0].custom_maintenance_user : null;

    if (!technician_user) {
        frappe.msgprint({
            title: __("Missing Technician"),
            indicator: "red",
            message: __("Please set <b>Maintenance User</b> in the Purposes table before creating Material Transfer.")
        });
        return;
    }

    // Look up employee warehouse by user
    frappe.db.get_list("Warehouse", {
        filters: { custom_user: technician_user },
        fields: ["name"],
        limit: 1
    }).then(wh_res => {
        const employee_wh = wh_res && wh_res[0] ? wh_res[0].name : "";

        if (!employee_wh) {
            frappe.msgprint({
                title: __("Warehouse Not Found"),
                indicator: "red",
                message: __(`No warehouse found for user <b>${technician_user}</b>. Please set a warehouse linked to this user.`)
            });
            return;
        }

        // Build items from Purposes (item_code + refill_qty)
        const items = [];
        purposes.forEach(row => {
            if (row.item_code && (row.custom_refill_qty || 0) > 0) {
                items.push({
                    item_code:   row.item_code,
                    qty:         row.custom_refill_qty,
                    s_warehouse: frm.doc.custom_from_warehouse,
                    t_warehouse: employee_wh
                });
            }
        });

        if (!items.length) {
            frappe.msgprint({
                title: __("No Items"),
                indicator: "orange",
                message: __("No items with <b>Refill Qty > 0</b> found in Purposes. Please fill Refill Qty first.")
            });
            return;
        }

        // Derive company from the from_warehouse (may differ from MV company)
        frappe.db.get_value("Warehouse", frm.doc.custom_from_warehouse, "company").then(r => {
            const se_company = (r && r.message && r.message.company) || frm.doc.company;

            frappe.new_doc("Stock Entry", {}, doc => {
                doc.stock_entry_type         = "Material Transfer";
                doc.company                  = se_company;
                doc.from_warehouse           = frm.doc.custom_from_warehouse;
                doc.to_warehouse             = employee_wh;
                doc.custom_maintenance_visit = frm.doc.name;

                doc.items = [];
                items.forEach(item => {
                    let row               = frappe.model.add_child(doc, "Stock Entry Detail", "items");
                    row.item_code         = item.item_code;
                    row.qty               = item.qty;
                    row.s_warehouse       = item.s_warehouse;
                    row.t_warehouse       = item.t_warehouse;
                    row.conversion_factor = 1;
                    row.transfer_qty      = item.qty;
                });

                frappe.set_route("Form", "Stock Entry", doc.name);
            });
        });
    });
}
